File: behavior/changepointprobability.m
function CPP = changepointprobability (input, hazardrate)
% changepointprobability %%
% This function was adapted from MR Nassar code (getOptimalLRs.m)to get
% optimal inferences over the probability of a binary variable (input) that
% evolves according to a change-point process with a prior over
% changepoints (hazardrate).
%
% Ref paper: Wilson, R. C., Nassar, M. R., & Gold, J. I. (2013). A
% Delta-rule approximation to Bayesian inference in change-point problems.
% PLoS Computational Biology, 9(7), e1003150

%AUTHORS:   H Atilgan 200417
%
%INPUT ARGUMENTS
%    input       : binary outcomes
%    hazardrate  : hazard rate

%OUTPUT ARGUMENTS
%   CPP: change-point probability
%%

% Check data vector - excludes nans (inlude them at the end):
data = input(~isnan(input));

% Create an grid for possible probability:
ps=(0:.01:1)';
cpPrior=ones(size(ps))./length(ps);
p=cpPrior;

% Calculate log likelihood based on actions:
dataLL=nan(size(data));  % Log likelihood
for i = 1:length(data)-1
    if ~isnan(data(i))
        if data(i)
            condProb=p.*ps;
        else
            condProb=p.*(1-ps);
        end
        dataLL(i+1)=log(sum(condProb)); % changep point probability for next trial
        % change-point probability - measure of suprise of switch
        p=p.*(1-hazardrate)+cpPrior.*(hazardrate);
        p=p./nansum(p); % normalize;
        
        % account for likelihood of data:
        if data(i)
            p=p.*ps;
        else
            p=p.*(1-ps);
        end
        p=p./nansum(p);
    end
end

% Calculate change-point probability based on data likelihood and
% true hazard rate:
Q = (.5.*hazardrate)./( exp(dataLL).*(1-hazardrate));
temp_CPP = Q./(1+Q);

CPP = nan(size(input)); % put nans back.
CPP(~isnan(input)) = temp_CPP;

File: reversalExpectation_py/behavior/changepoint_probability.py
"""
changepoint_probability.py
==========================
Translation of changepointprobability.m
(H Atilgan 200417, adapted from MR Nassar / getOptimalLRs.m).

Computes the trial-by-trial change-point probability (CPP) for a
binary outcome sequence under a change-point prior.

Reference:
    Wilson, R.C., Nassar, M.R., & Gold, J.I. (2013).
    A delta-rule approximation to Bayesian inference in change-point
    problems. PLoS Computational Biology, 9(7), e1003150.
"""

import numpy as np


def changepoint_probability(
    input_seq: np.ndarray,
    hazardrate: float,
) -> np.ndarray:
    """
    Translation of changepointprobability.m.

    Parameters
    ----------
    input_seq  : float array
        Binary outcome sequence (0 / 1). NaN values are ignored during
        the computation and reinserted in the output at their original
        positions.
    hazardrate : float
        Prior probability of a change-point at each trial (0 < H < 1).

    Returns
    -------
    CPP : float array (same length as input_seq)
        Trial-by-trial change-point probability. NaN where input is NaN.
    """
    input_seq = np.asarray(input_seq, dtype=float)
    data      = input_seq[~np.isnan(input_seq)]
    m         = len(data)

    # Grid over possible outcome probabilities
    ps       = np.linspace(0, 1, 101)           # 0.00, 0.01, … 1.00
    cp_prior = np.ones(len(ps)) / len(ps)
    p        = cp_prior.copy()

    data_ll = np.full(m, np.nan)

    for i in range(m - 1):
        if data[i]:
            cond_prob = p * ps
        else:
            cond_prob = p * (1 - ps)

        data_ll[i + 1] = np.log(np.sum(cond_prob))

        # Change-point update
        p = p * (1 - hazardrate) + cp_prior * hazardrate
        p = p / np.nansum(p)

        # Likelihood update
        if data[i]:
            p = p * ps
        else:
            p = p * (1 - ps)
        p = p / np.nansum(p)

    # CPP from data log-likelihood and hazard rate
    Q        = (0.5 * hazardrate) / (np.exp(data_ll) * (1 - hazardrate))
    temp_cpp = Q / (1 + Q)

    # Reinsert NaNs at original positions
    CPP = np.full(len(input_seq), np.nan)
    CPP[~np.isnan(input_seq)] = temp_cpp

    return CPP
